lib/LANraragi/Utils/Routing.pm

```diff
@@ -150,6 +152,16 @@ sub apply_routes {
     $logged_in_api->put('/api/categories/:id/:archive')->to('api-category#add_to_category');
     $logged_in_api->delete('/api/categories/:id/:archive')->to('api-category#remove_from_category');
 
+    # Reading Group API
```

Comment:
Yes, I left all public for the Demo to be easily testeable, but I'll change all the POST, UPDATE, and DELETE  EPs to login required when I commit the final version.

I can make an EP to return the groups where an Arcid is present, only that complexity could raise very quickly depending in how is used. For example, my main idea(As someone that hasn't worked a lot with Redis, so maybe I'm saying gibberish) was to create an exclusive table GROUPED which would contain a list of hashes, where a hash would contain a list of the group ids where is contained, so this way: Want to know if the work is contained somewhere? Check if the Arcid is in the list of hashes, Want to know which lists is contained? Access the key and that's all, that way the search would be, probably at most, O(n), otherwise to search in every list and between all the works would go to O(n*m). I'll check what can be done with Redis and update with my findings.

Reply:
Currently the equivalent endpoint for categories just iterates over every category in the DB and checks if the arcid is present in them. Its as inefficient as can be but I haven't received any complaints about it!  

Tanks could however have a simpler database structure than categories since they're just a name and a sorted list of IDs. (whereas categories have to bundle a search predicate, as well as pinned/last_used status)  

You might be able to make use of [Redis sorted sets](https://redis.io/docs/data-types/sorted-sets/) to save the ordered list of arcids (as `TANK_xxxxx` with the name in a separate `TANK_xxxxx_NAME` I guess?), at which point you could use stuff like `ZSCORE` to find the ID in each set, with each zscore being a O(1) operation.

Reply:
Ok, I've been testing the sorted sets for a while to check how to implement it, and seems doable, but there is a pair of problems(Which mostly are not something big, but are worth to mention):

First is that 'zrangebyscore' from Perl seems to not be doing what is supposed
![image](https://github.com/Difegue/LANraragi/assets/5926318/a54eb37f-c226-4359-8d37-7ad6173ed1f6)
Here in the first print and 3rd print you can see a call from this exact line `my %tankoubon = $redis->zrangebyscore($tank_id, 0, "+inf", "WITHSCORES");` but in one it returned the results ordered, and in the other reversed, this might be a problem, because it makes me require to add an aditional sort to make sure the list is ordered. Is not that big of a problem, but performance wise might eventually take a hit(Although, with this I can add pagination more easily with the LIMIT offset count option, so maybe is even better). BTW score 0 will be reserved for the name of the Tank, so no need to save it in the hash key.

Second is that, because I can't just overwrite what is saved in one, reorder and edit will probably require a little more work, since now I'll need to check which arcids are not more in what Front sends me to remove them with zrem, and manually update all the scores with zadd. I'm not sure if this could be a bad practice, but at the very least in the docs it says that those methods are O(log(N)) and O(M*log(N)), and I doubt someone will create a list with tens of thousands of files as to be notable, so I'm guessing it's ok. 

Will try to keep the EPs working the same, but maybe something that will change is that I'll add pagination in the list of Tanks and the list of archives, just to test.

Reply:
Looking back at this comment -- Do you still have the first problem? zrange and friends shouldn't reverse the order every time you call them!  

As for the second one, imo we could change the API to just have add/delete endpoints (with an optional position argument for add) and hand off all the complexity of knowing which archives to add/remove to the frontend/clients... 
The issue with that is that I don't think there's an easy way to update the scores of every ID in a sorted set when we move/insert a new one 🤔

Reply:
Hi, sorry for the lack of answers, I'm in the middle of a vacation, on Monday that I return I'll check the changes you requested, but for the moment I'll answer this one:

The 1st problem still happens, but as mentioned I solved it by manually sorting after the zrangebyscore, I'm not sure why is it giving me unordered data, but it happens at least on my machine, if someone can verify that is working correctly, then the loop marked with `# Sort and add IDs to archives array` in get_tankoubon can be removed(With the required changes to make the next lines keep working ofc).

For the 2nd, I already implemented those 2 methods in the update to make sure to apply all the insert/deletes in one in the update_archive_list with these lines

![image](https://github.com/Difegue/LANraragi/assets/5926318/9ab6bbf1-2ba5-4af1-a269-b385fc4a59e5)

Basically, what I do is to check the left difference with the original, and arcids not anymore are removed with zrem, while with the new list I just regenerate all the scores(Again, for this I'm taking the array that is sent from Frontend is necessarily ordered) and update it that way. Simple and it works.

The problem with having add/delete endpoints is that if you want to add it at an specific place, you need to regenerate all the scores anyway, for example, if you have a 100 elements list, and you want to insert in position 4, from the actual 4 onward you would need to add 1 to the score and update, same case with the deletes, otherwise the list might end up messed up eventually with dupped scores. I still added an add_to_tankoubon method, but it adds it at the final position of the tank, since I can use zcard to get the size of the set to know its score. 

Also, just to reminder that another of the problems is the number of requests you would need to do to the back, might not be very bad, but sometimes might be better 1 good request than 100 smaller ones.

Reply:
No real rush! I'm going to be away next week as well so you can take your time. 👍  

I'll try removing the loop when running this locally at some point to verify - although if it's out of order on _some machines_ we'd better keep the sorting code in anyway..
I agree with the score regen being an unfixable problem (or at least idk how to fix it), so if we need to rescore everything might as well keep the current approach.